src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -1724,109 +1870,505 @@ ep_rt_mono_write_event_method_load (
+		g_free (exception_message);
+		g_free (type_name);
+	}
+
+	return true;
+}
+
+bool
+ep_rt_mono_write_event_exception_clause (
+	MonoMethod *method,
+	uint32_t clause_num,
+	MonoExceptionEnum clause_type,
+	MonoObject *obj)
+{
+	if ((clause_type == MONO_EXCEPTION_CLAUSE_FAULT || clause_type == MONO_EXCEPTION_CLAUSE_NONE) && (!EventEnabledExceptionCatchStart() || !EventEnabledExceptionCatchStop()))
+		return true;
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FILTER && (!EventEnabledExceptionFilterStart() || !EventEnabledExceptionFilterStop()))
+		return true;
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FINALLY && (!EventEnabledExceptionFinallyStart() || !EventEnabledExceptionFinallyStop()))
+		return true;
+
+	uintptr_t ip = 0; //TODO: Have profiler pass along IP of handler block.
+	uint64_t method_id = (uint64_t)method;
+	char *method_name = NULL;
+
+	method_name = mono_method_get_name_full (method, TRUE, TRUE, MONO_TYPE_NAME_FORMAT_IL);
+
+	if ((clause_type == MONO_EXCEPTION_CLAUSE_FAULT || clause_type == MONO_EXCEPTION_CLAUSE_NONE)) {
+		FireEtwExceptionCatchStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionCatchStop (
+			NULL,
+			NULL);
+	}
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FILTER) {
+		FireEtwExceptionFilterStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionFilterStop (
+			NULL,
+			NULL);
+	}
+
+	if (clause_type == MONO_EXCEPTION_CLAUSE_FINALLY) {
+		FireEtwExceptionFinallyStart (
+			(uint64_t)ip,
+			method_id,
+			(const ep_char8_t *)method_name,
+			clr_instance_get_id (),
+			NULL,
+			NULL);
+
+		FireEtwExceptionFinallyStop (
+			NULL,
+			NULL);
+	}
+
+	g_free (method_name);
+	return true;
+}
+
+bool
+ep_rt_mono_write_event_monitor_contention_start (MonoObject *obj)
+{
+	if (!EventEnabledContentionStart_V1 ())
+		return true;
+
+	FireEtwContentionStart_V1 (
+		0 /* ManagedContention */,
+		clr_instance_get_id (),
+		NULL,
+		NULL);
+
+	return true;
+}
+
+bool
+ep_rt_mono_write_event_monitor_contention_stop (MonoObject *obj)
+{
+	if (!EventEnabledContentionStop ())
+		return true;
+
+	FireEtwContentionStop (
+		0 /* ManagedContention */,
+		clr_instance_get_id (),
+		NULL,
+		NULL);
+
+	return true;
+}
+
+bool
+ep_rt_mono_write_event_method_jit_memory_allocated_for_code (
+	const uint8_t *buffer,
+	uint64_t size,
+	MonoProfilerCodeBufferType type,
+	const void *data)
+{
+	if (!EventEnabledMethodJitMemoryAllocatedForCode ())
+		return true;
+
+	if (type != MONO_PROFILER_CODE_BUFFER_METHOD)
+		return true;
+
+	uint64_t method_id = 0;
+	uint64_t module_id = 0;
+
+	if (data) {
+		MonoMethod *method;
+		method = (MonoMethod *)data;
+		method_id = (uint64_t)method;
+		if (method->klass)
+			module_id = (uint64_t)(uint64_t)m_class_get_image (method->klass);
 	}
 
+	FireEtwMethodJitMemoryAllocatedForCode (
+		method_id,
+		module_id,
```

Comment:
Why do some of these events work with method ids, and some with full method names (for example in `ep_rt_mono_write_event_exception_clause`)? synchronously stringifying method names on every exception seems like it will add up.

Reply:
Agree, this is a direct result of how the underlying native events have been defined and not something we can change.